src/Infrastructure/IO/src/ESMCI_IO_Handler.C

```diff
@@ -361,6 +366,101 @@ void IO_Handler::finalize (
 
+//-----------------------------------------------------------------------------
+#undef  ESMC_METHOD
+#define ESMC_METHOD "ESMCI::IO_Handler::arrayRead()"
+//BOPI
+// !IROUTINE:  ESMCI::IO_Handler::arrayRead    - Read an array from a file
+//
+// !INTERFACE:
+void IO_Handler::arrayRead(
+//
+// !RETURN VALUE:
+//
+//
+// !ARGUMENTS:
+//
+  Array *arr_p,                           // (inout) - Destination of read
+  const char * const name,                // (in)    - Optional array name
+  int *timeslice,                         // (in)    - Optional timeslice
+  int *rc                                 // (out)   - Error return code
+  ) {
+//
+// !DESCRIPTION:
+//    Read data from field <name> from the open file. If timeslice is not
+//    NULL, it should point to an integer representing the timeslice to read
+//    from the Array.
+//
+//EOPI
+//-----------------------------------------------------------------------------
+  // initialize return code; assume routine not implemented
+  int localrc = ESMF_RC_NOT_IMPL;         // local return code
+  if (rc != NULL) {
+    *rc = ESMF_RC_NOT_IMPL;               // final return code
+  }
+
+  PRINTPOS;
+
+  for (int tile = 1; tile <= ntiles; ++tile) {
+    arrayReadOneTileFile(arr_p, tile, name, timeslice, &localrc);
+    if (ESMC_LogDefault.MsgFoundError(localrc, ESMCI_ERR_PASSTHRU, ESMC_CONTEXT, rc))
+      return;
+  }
+  // return successfully
+  if (rc != NULL) {
+    *rc = ESMF_SUCCESS;
+  }
+} // IO_Handler::arrayRead()
+//-----------------------------------------------------------------------------
+
+//-----------------------------------------------------------------------------
+#undef  ESMC_METHOD
+#define ESMC_METHOD "ESMCI::IO_Handler::arrayWrite()"
+//BOPI
+// !IROUTINE:  ESMCI::IO_Handler::arrayWrite    - Write an Array to a file
+//
+// !INTERFACE:
+void IO_Handler::arrayWrite(
+//
+// !RETURN VALUE:
+//
+//
+// !ARGUMENTS:
+  Array *arr_p,                           // (in) Destination of write
+  const char * const name,                // (in) Optional array name
+  const std::vector<std::string> &dimLabels, // (in) Optional dimension labels
+  int *timeslice,                         // (in) Optional timeslice
+  const ESMCI::Info *varAttPack,            // (in) Optional per-variable Attribute Package
+  const ESMCI::Info *gblAttPack,            // (in) Optional global Attribute Package
+  int *rc                                 // (out) - Error return code
+//
+  ) {
+//
+// !DESCRIPTION:
+//    Write data to field <name> to the open file.
+//
+//EOPI
+//-----------------------------------------------------------------------------
+  // initialize return code; assume routine not implemented
+  int localrc = ESMF_RC_NOT_IMPL;         // local return code
+  if (rc != NULL) {
+    *rc = ESMF_RC_NOT_IMPL;               // final return code
+  }
+
+  PRINTPOS;
+
+  for (int tile = 1; tile <= ntiles; ++tile) {
```

Comment:
@billsacks for this loop (and similarly for reads) are we getting any concurrency on the write when tiles are on completely disjoint sets of procs?

Reply:
> for this loop (and similarly for reads) are we getting any concurrency on the write when tiles are on completely disjoint sets of procs?

Unfortunately, no. Getting concurrency would require a more substantial rework of the PIO interface so that we have separate PIO instances for each tile. From our earlier discussions, we decided not to do that for the first iteration.